kernel/src/thread/task.rs

```diff
@@ -88,15 +90,71 @@ pub fn create_new_user_task(
             if current_thread.is_exited() {
                 break;
             }
-            handle_pending_signal(user_ctx, &ctx, syscall_number);
-            // If current is suspended, wait for a signal to wake up self
-            while current_thread.is_stopped() {
-                Thread::yield_now();
-                debug!("{} is suspended.", current_posix_thread.tid());
-                handle_pending_signal(user_ctx, &ctx, None);
+
+            enum Status {
+                /// The thread has exited.
+                Exit,
+                /// The thread is currently running.
+                Run,
+                /// The thread is stopped, but there are pending signals.
+                /// We need to handle the pending signals to determine if any can
+                /// wake up the thread.
+                RecheckSignal,
             }
-            if current_thread.is_exited() {
-                debug!("exit due to signal");
+
+            let mut status = Status::Run;
+
+            loop {
+                // Fast path: Check if the thread has exited or is not stopped
```

Comment:
`WaitQueue::pause_until` already has a fast path so there is no need to have another.

(The comment above suggests to use `Waiter::pause_until`, which does not have a fast path, though.)